judoscale/celery/tasks_monitor.py

```diff
@@ -0,0 +1,87 @@
+import logging
+from datetime import datetime as dt
+#from celery import Celery
+from celery import current_app
+
+from judoscale.core.metric import Metric
+from judoscale.core.reporter import reporter
+from judoscale.core.metrics_store import metrics_store
+
+logger = logging.getLogger(__name__)
+
+
+class EventLogger:
+    def log_task_status_change(self, task, event):
+        print('[{}] {} {} (STATE={}, UUID={})'.format(
+            self._to_datetime(task.timestamp),
+            event['type'].upper(),
+            task.name,
+            task.state.upper(),
+            task.uuid
+        ))
+    def _to_datetime(self, timestamp):
+        return dt.fromtimestamp(timestamp) if timestamp is not None else None
+
+
+class CeleryEventsHandler:
+    def __init__(self, celery_app, verbose_logging=False):
+        self._app = celery_app
+        self._state = celery_app.events.State()
+        self._logger = EventLogger()
+        self._verbose_logging = verbose_logging
+        self.task_started = {}
+        self.task_created = {}
+
+    def announce_task_received(self, event):
+        self._state.event(event)
+        task = self._state.tasks.get(event["uuid"])
+        self.task_created[task.uuid] = task.timestamp
```

Comment:
I'm really struggling to follow what's happening in this method. What is `task`, and how is it different from `event`? Where does `task.timestamp` come from? Why are we adding the timestamp to `task_created`, only to read from it again a few lines later?

Reply:
These methods are suggested by Celery's monitoring:
https://docs.celeryq.dev/en/stable/userguide/monitoring.html#events
Events are messages sent to the worker related to the tasks. We are adding the timestamps in the CeleryEventHandler class attribute so the other methods can read from these. I was using the time lag between task_created and task_started before for the metrics, and also in the method task_succeeded to check how long it took for the task to complete. I put it back now, so you understand better the reasoning.

Reply:
see https://github.com/rails-autoscale/judoscale-python/pull/19/files#diff-41f56399c272d79f41844d1600335d8c402315b04b982485aef065d8d967d6a0R72